aptos-move/move-examples/marketplace/sources/token_offer.move

```diff
@@ -0,0 +1,788 @@
+        let token_offer_addr = object::object_address(&token_offer);
+        assert!(
+            exists<TokenOfferTokenV2>(token_offer_addr),
+            error::not_found(ENO_TOKEN_OFFER),
+        );
+
+        // Check it's the correct token
+        let seller_address = signer::address_of(seller);
+        let token = borrow_global<TokenOfferTokenV2>(token_offer_addr).token;
+        assert!(seller_address == object::owner(token), error::permission_denied(ENOT_TOKEN_OWNER));
+
+        // Move the token to its destination
+        let recipient = object::owner(token_offer);
+        object::transfer(seller, token, recipient);
+
+        // Pay fees
+
+        let royalty = tokenv2::royalty(token);
+        let (royalty_payee, royalty_denominator, royalty_numerator) = if (option::is_some(&royalty)) {
+            let royalty = option::destroy_some(royalty);
+            let payee_address = royalty::payee_address(&royalty);
+            let denominator = royalty::denominator(&royalty);
+            let numerator = royalty::numerator(&royalty);
+            (payee_address, denominator, numerator)
+        } else {
+            (signer::address_of(seller), 1, 0)
+        };
+
+        settle_payments<CoinType>(
+            object::owner(token_offer),
+            seller_address,
+            token_offer_addr,
+            royalty_payee,
+            royalty_denominator,
+            royalty_numerator,
+            events::token_metadata_for_tokenv2(token),
+        );
+    }
+
+    /// From the coin offer remove appropriate payment for the token and distribute to the seller,
+    /// the creator for royalties, and the marketplace for commission. If there are no more slots,
+    /// cleanup the offer.
+    inline fun settle_payments<CoinType>(
+        buyer: address,
+        seller: address,
+        token_offer_addr: address,
+        royalty_payee: address,
+        royalty_denominator: u64,
+        royalty_numerator: u64,
+        token_metadata: events::TokenMetadata,
+    ) acquires CoinOffer, TokenOffer, TokenOfferTokenV1, TokenOfferTokenV2 {
+        assert!(exists<TokenOffer>(token_offer_addr), error::not_found(ENO_TOKEN_OFFER));
+        let token_offer_obj = borrow_global_mut<TokenOffer>(token_offer_addr);
+        assert!(
+            timestamp::now_seconds() < token_offer_obj.expiration_time,
+            error::invalid_state(EEXPIRED),
+        );
+        let price = token_offer_obj.item_price;
+
+        assert!(
+            exists<CoinOffer<CoinType>>(token_offer_addr),
+            error::not_found(ENO_COIN_OFFER),
+        );
+        let coin_offer = borrow_global_mut<CoinOffer<CoinType>>(token_offer_addr);
+        let coins = coin::extract(&mut coin_offer.coins, price);
+
+        let royalty_charge = price * royalty_numerator / royalty_denominator;
+        let royalties = coin::extract(&mut coins, royalty_charge);
+        coin::deposit(royalty_payee, royalties);
```

Comment:
This one is a bit harder because you will possibly never have the royalty signer, but if it isn't registered for `CoinType` this function will always fail, so maybe you need to just assert `is_account_registered<CoinType>(royalty_payee)` when the offer is initially created..?

Reply:
Easy enough, all now register accordingly